app/models/piece.rb

```diff
@@ -104,72 +102,88 @@ def legal_move?(new_x, new_y) # used only when checking for stalemate in a parti
-    vertical_obstruction?(range_y) || horizontal_obstruction?(range_x)
+  def trimmed_coordinates_array(x_values, y_values) # x_values = [1, 1, 1, 1], y_values = [2, 3, 4, 5]
+    coordinates_array = x_values.zip(y_values) # [[1, 2], [1, 3], [1, 4], [1, 5]]
+    coordinates_array.shift
+    coordinates_array.pop
+    coordinates_array # [[1, 3], [1, 4]]
   end
 
-  def diagonal_obstruction?(new_x, new_y)
-    return false unless diagonal?(new_x, new_y)
+  def straight_obstruction_array(new_x, new_y)
+    return nil unless straight_move?(new_x, new_y)
     x_values = if x_position.to_i < new_x.to_i
                  (x_position.to_i..new_x.to_i).to_a # array of x values, including the starting and ending squares
-               else
+               elsif x_position.to_i > new_x.to_i
                  x_position.to_i.downto(new_x.to_i).to_a
+               else # new_x must equal x_position in this last case. and we already ruled out range_y != 0 in straight_move?
+                 range_y = (new_y - y_position).abs
+                 Array.new(range_y + 1, new_x.to_i) # array of x values that are all equal to new_x, the length of which is the same as the length of y values array
                end
 
     y_values = if y_position.to_i < new_y.to_i
                  (y_position.to_i..new_y.to_i).to_a # array of y values, including the starting and ending squares
-               else
+               elsif y_position.to_i > new_y
                  y_position.to_i.downto(new_y.to_i).to_a
+               else # new_y must equal y_position in this last case. and we already ruled out range_x != 0 in straight_move?
+                 range_x = (new_x - x_position).abs
+                 Array.new(range_x + 1, new_y.to_i)
                end
+    trimmed_coordinates_array(x_values, y_values)
+  end
 
-    coordinates = []
-
-    x_values.zip(y_values) do |x_value, y_value| # combines x_value and y_value pairs into a multidimensional array
-      coordinates << [x_value, y_value] # [ [1, 1], [2, 2], [3, 3] ] for example
+  def can_be_blocked?(new_x, new_y) # used for determining checkmate. Our piece is the attacking piece in this case.
+    # this method works for all pieces except a knight, which doesn't have a straight path.
+    game.pieces.where(is_white: !is_white).where.not(x_position: nil, y_position: nil, type: KING).find_each do |piece|
+      straight_obstruction_array(new_x, new_y).each do |coords|
+        return true if piece.valid_move?(coords.first, coords.last)
+      end
     end
+  end
 
-    coordinates.shift
-    coordinates.pop # this is an array of only the in-between squares - not including the start or end squares
-    coordinates.each do |coor|
-      obstructing_piece = game.get_piece_at_coor(coor.first, coor.last)
-      return true if obstructing_piece.present?
-    end
-    false
+  def causes_check?(is_white)
+    game.under_attack?(!is_white, enemy_king(is_white).x_position, enemy_king(is_white).y_position)
```

Comment:
Okay, I'm going to try to explain this one. Honestly, it took me a while to work this out, and if I don't explain myself well here, I'm happy to hop on a call with any of you to try to explain in a verbal conversation.

You're right that causes_check? is meant to be called on the attacking piece, which has the opposite is_white value from its enemy_king.

!is_white is basically taking the is_white parameter which was passed into the causes_check? method and saying: 
for the piece causing check, take its is_white value.
The is_white value of the king that piece is attacking will be !is_white.

game.under_attack? is a method that is about a specific square. You need to pass in both the color (the is_white value) of the piece on that square, as well as the coordinates of that square. Then you can determine if a piece of the particular color you passed in is under attack by any pieces of the opposite team.

What might also be confusing is that enemy_king(is_white) is the enemy king of the attacking piece instance, and it accepts the is_white value of the attacking piece. enemy_king(is_white)'s is_white value is !is_white. Does that clear anything up at all?

I think my code works, but I know it is a bit hard to read, and I'm definitely open to refactoring. Let me know your thoughts.

Reply:
I'm going to try again with your example, Dennis.

Let's say a white bishop moves and it places the black king in check by its move.
causes_check? should be called on that white bishop's move to see if that move will place the black king in check.
The black king being in check is a specific instance of the black king's square being under attack.
(I thought it was helpful to separate out the under_attack method so we could also look and see if the attacking_piece's square was under_attack.)
For a square to be under_attack, we look to see if any pieces of the other color could move to that square in a valid_move.
For the black king's square, we look to see if any white pieces could move to that square. (We "filter for pieces that have is_white value of true when checking to see which pieces could capture the" black king "at the given x and y" - as you said above.)
It may seem like circular reasoning at first - to define causes_check (for a white piece) by seeing if the black king's square is under_attack by a white piece... but it seems to work.

This will also catch instances where, let's say a white pawn was in between the white bishop and the black king. Then the white pawn moved and exposed the black king to attack by the white bishop. The white pawn then caused check, because it caused the black king's square to be under_attack, _even though it wasn't actually the pawn putting the king in check!_

Reply:
But. This exposes a big flaw in my logic. The attacking_piece can't actually be defined as being the piece that moved last, which is how I have it set up right now. How else can we define the attacking_piece? Or do we need to re-think all of this?

Reply:
I would loop through each of the pieces on the side of the piece that was last moved and return the first piece that can attack the square the opposing king is on.  Also, you could use enemy_king(is_white).is_white as the first argument for game.under_attack, and put some comments in both the causes_check and under_attack methods to make it clear what each of the arguments should be.